### DISCUSSION TITLE
Post-SciPy 2024 tutorial notes

### DISCUSSION THREAD

**Discussion by jonmmease** [vega/altair#3476] [General]
@ChristopherDavisUCI and I gave a Vega-Altair tutorial at SciPy 2024 Monday morning. The tutorial materials are at https://github.com/vega/SciPy2024-Altair-Tutorial.  I just wanted to record some notes on potential Vega-Altair features / fixes that participants brought up.

 1. **expressions are hard to debug**: On participant ran into a situation when using the `filter` transform where they misspelled one of the columns in a `datum.{col}` expressions. This doesn't result in an error, as Vega (and VegaFusion) follow JavaScript's semantics where this expressions returns `undefined` rather than erroring. This resulted in the entire filter expression coercing to `false`, and all of the data being dropped.  Making this an error in Vega would be a breaking change, but I think we could make VegaFusion more strict about this.
 2.  **better type safety in aggregation expressions**: On participant asked if it would be possible for the type checker to catch errors in the aggregation string in an expression like `alt.Y("countt()")`. I didn't know off hand whether there's a more general form possible here that would type check. Or perhaps we could write a [mypy plugin](https://mypy.readthedocs.io/en/stable/extending_mypy.html) that would validate these aggregation strings.
 3. **desire to scale size of interacive charts**: One participant asked if it were possible to scale up the physical size of charts. This is possible using the `png` renderer by setting `scale_factor` to 2+, but I don't think it's possible to accomplish this with an interactive chart.  This would need to be implemented in Vega if it's not already possible.
 4. **support PIL images**: One participant wanted to make a chart with an image mark [like this](https://altair-viz.github.io/user_guide/marks/image.html#scatter-plot-with-image-marks), but their images were stored locally as [PIL](https://pillow.readthedocs.io/en/stable/) Images.  I explained that it's possible to convert PIL images to base64 encoded PNG urls, but this is a bit of an annoying process. In plotly, I added support for doing this automatically during serialization and I think we could do the same thing in Vega-Altiar. Here's the [`pil_image_to_uri` utility function](https://github.com/plotly/plotly.py/blob/241746e203625544a22e7d667460474bc45b413c/packages/python/plotly/_plotly_utils/basevalidators.py#L2423-L2434) I added in plotly for this conversion.
 5. **self-hosted Vega editor**: Someone asked if it's possible to host the Vega editor yourself. And if so, whether it's possible to point Altair's `to_url()` and `open_editor()` methods to this instance.  I'm thinking it could make a lot of sense to have Altair check for a `VEGA_EDITOR_URL` environment variable, and it found use this for these methods. I'm also not sure if it's currently possible to override the editor URL in the `Open in Editor` dropdown entry.

I can turn these into issues later, but just wanted to document the notes quickly while they were fresh in my mind.

**Comment by joelostblom** [comment]
Thanks for sharing the material and some of the takeaways from Scipy!

>  Making this an error in Vega would be a breaking change

Although it is breaking, it sounds quite useful to receive an error instead of the expression silently not doing what's expected, so maybe it could be considered for a future major Vega release?

**Comment by dangotbanned** [comment]
> 2. **better type safety in aggregation expressions**: 
> On participant asked if it would be possible for the type checker to catch errors in the aggregation string in an expression like `alt.Y("countt()")`. 
> I didn't know off hand whether there's a more general form possible here that would type check. 
> Or perhaps we could write a [mypy plugin](https://mypy.readthedocs.io/en/stable/extending_mypy.html) that would validate these aggregation strings.

@jonmmease I was holding off on this until an issue was opened, but I've rolled it into a longer-term *RFC* as the `agg` class in #3505.

I thought it might be helpful sharing here how I arrived there, as it is an alternative solution for the same problem.

---

# Why not a [mypy plugin](https://mypy.readthedocs.io/en/stable/extending_mypy.html)?
There are several other type checkers [spec](https://typing.readthedocs.io/en/latest/#typing-related-tools), [discussion](https://www.reddit.com/r/Python/comments/1ajnikt/to_pycharm_users_how_are_you_type_checking_your/) and this would only work for `mypy` users.
I'm also unsure how an optional dependency would work in annotations. Especially since they would no longer be `str`.

# If not `mypy` then what?
I think [`typing.Annotated`](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#annotated)([PEP 593](https://peps.python.org/pep-0593/)) would be something to explore.

Some examples from popular libraries:
- [msgspec](https://jcristharif.com/msgspec/constraints.html#string-constraints)
- [pydantic](https://docs.pydantic.dev/latest/concepts/fields/#string-constraints)

So if we say `Annotated` works for this, the next issue is what does such an annotation actually look like?

## Existing code

<details>
<summary>Building up the patterns</summary>

https://github.com/vega/altair/blob/679a7ce30cb3366319bce2f4211dd62a5a80c87c/altair/utils/core.py#L196-L204

</details>

<details>
<summary>Inference, additionally factoring in the `Chart.data` </summary>

https://github.com/vega/altair/blob/679a7ce30cb3366319bce2f4211dd62a5a80c87c/altair/utils/core.py#L503-L687

</details>


## Thoughts
If `parse_shorthand` used a simple regex pattern, I would be more in favour of exploring this.

Not saying it couldn't be solved, but what we have is pretty complex and somewhat obfuscated. 

# Alternative solution
- #3505

I approached this quite differently, but it aligns well with upcoming changes to `expr` from #3466 and proposed changes in #3239 

## Examples
Aggregation functions are `classmethod`s accessible under `agg`.
This simplifies the problem as incorrect method names are a runtime error:

![image](https://github.com/user-attachments/assets/d2c64c4f-a49e-413c-8f1d-8a192fa13bc5)

These all have the same 2 arguments, both of which are **optional**.
Each method docstring *could* also provide examples on usage, something that wouldn't be feasible before

![image](https://github.com/user-attachments/assets/9d9bb2a8-f934-4fc4-a4d9-d8e18ea1931b)

Auto-complete on the `type` argument

![image](https://github.com/user-attachments/assets/7dd2c173-30c5-420f-96cb-c5f9eb4e561b)

Both static **AND** runtime checking, so even **non-typing users** get some benefits:

![image](https://github.com/user-attachments/assets/d14fea4a-3932-495c-9b46-adee8d198db3)


A range of examples:

![image](https://github.com/user-attachments/assets/c7ae3149-1206-44f6-95c0-c9c0bac6880d)

Backwards-compatible string syntax via `alt.agg()` like `alt.expr()`:

![image](https://github.com/user-attachments/assets/102e65b5-dd01-4255-a376-680839ca7762)

Integrating this was an incredibly minor change in [`5929a5e` (#3505)](https://github.com/vega/altair/pull/3505/commits/5929a5e550f8747a296c801fb1979fd59e1fa3db)

![image](https://github.com/user-attachments/assets/acfec804-ba99-4700-b81b-c3bd645f4d0d)

**Reply by jonmmease** [reply]
Thanks for sharing @dangotbanned, this looks like a very nice approach, and consistent with the expression language. I'm in favor!
